src/Microsoft.ML.Transforms/Text/CharTokenizeTransform.cs

```diff
@@ -438,6 +442,67 @@ private ValueGetter<VBuffer<ushort>> MakeGetterVec(IRow input, int iinfo)
 
                     dst = new VBuffer<ushort>(len, values, dst.Indices);
                 };
+
+            ValueGetter < VBuffer<ushort> > valueGetterCurrentVersion = (ref VBuffer<ushort> dst) =>
+                {
+                    getSrc(ref src);
+
+                    int len = 0;
+
+                    for (int i = 0; i < src.Count; i++)
+                    {
+                        if (src.Values[i].HasChars)
+                        {
+                            len += src.Values[i].Length;
+
+                            if (i > 0)
+                                len += 1;  // add UnitSeparator character to len that will be added
+                        }
+                    }
+
+                    if (_useMarkerChars)
+                        len += TextMarkersCount;
+
+                    var values = dst.Values;
+                    if (len > 0)
+                    {
+                        if (Utils.Size(values) < len)
+                            values = new ushort[len];
+
+                        int index = 0;
+
+                        // VBuffer<DvText> can be a result of either concatenating text columns together 
+                        // or application of word tokenizer before char tokenizer in TextTransform.
+                        //
+                        // Considering VBuffer<DvText> as a single text stream.
+                        // Therefore, prepend and append start and end markers only once i.e. at the start and at end of vector.
+                        // Insert UnitSeparator after every piece of text in the vector.
+                        if (_useMarkerChars)
+                            values[index++] = TextStartMarker;
+
+                        for (int i = 0; i < src.Count; i++)
+                        {
+                            if (!src.Values[i].HasChars)
+                                continue;
+
+                            if (i > 0)
+                                values[index++] = UnitSeparator;
+
+                            for (int ich = 0; ich < src.Values[i].Length; ich++)
+                            {
+                                values[index++] = src.Values[i][ich];
+                            }
+                        }
+
+                        if (_useMarkerChars)
+                            values[index++] = TextEndMarker;
+
+                        Contracts.Assert(index == len);
+                    }
+
+                    dst = new VBuffer<ushort>(len, values, dst.Indices);
+                };
+            return CurrentModelVersion < version.VerReadableCur ? valueGetterOldVersion : valueGetterCurrentVersion;
```

Comment:
@TomFinley and @Ivanidzo4ka, Is this correct way to maintain backward compatibility?

Reply:
Not exactly. For one thing you've stored this in a static variable, which is not correct, since this behavior will be specific to the instance. So if you read *two* instances of the transform, one with the old version, one with the new version, you'd break one or the other (depending on the order in which you read them). As a general rule, not just in ML.NET but in any coding anywhere, storing *instance* specific data inside a *static* variable is generally a bad idea. :)

For the sake of resaving an *old* model to the *new* format, you ought to store a boolean variable in the new  version of the file, to control whether it is the old or new version. So the saver will have `ctx.Writer.WriteBoolByte(_isStartEnd)` and in the reader have something like `_isStartEnd = ctx.Header.ModelVerReadable < 0x00010002 || ctx.Reader.ReadBoolByte()`. This hypothetical `_isStartEnd` will be `private readonly`, and emphatically not `public static`.

If anything about that is not clear let me know, thanks @zeahmed!

---
In reply to: [203896119](https://github.com/dotnet/machinelearning/pull/548#discussion_r203896119) [](ancestors = 203896119)